### PR TITLE
Layout animations cleanup

### PR DESCRIPTION
**Pull request by jwajgelt**
## Summary
Clean up the layout animation rewrite from #3332

- use Enum types for layout animation types ('entering', 'exiting' and 'layout') in both JS and native parts of the implementation
- try caching `AnimationsManager` in `LayoutAnimations.java` instead of calling `getNativeModule` (Android implementation)

### CODE REVIEW

File: Common/cpp/NativeModules/NativeReanimatedModule.cpp [hunk 1]
```diff
@@ -388,7 +388,7 @@ jsi::Value NativeReanimatedModule::configureLayoutAnimation(
     const jsi::Value &config) {
   layoutAnimationsManager_.configureAnimation(
       viewTag.asNumber(),
-      type.asString(rt).utf8(rt),
+      (LayoutAnimationType)type.asNumber(),
```
Comment (tomekzaw) [inline Common/cpp/NativeModules/NativeReanimatedModule.cpp:391]: C-style casts are not welcome, please use `static_cast` or define a new function `parseLayoutAnimationType` that accepts `const jsi::Value &`.

File: android/src/main/java/com/swmansion/reanimated/layoutReanimation/LayoutAnimations.java [hunk 2]
```diff
@@ -9,6 +9,13 @@
 import java.util.Map;
 
 public class LayoutAnimations {
+  public static class Types {
+    static final int ENTERING = 0;
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/layoutReanimation/LayoutAnimations.java:13]: Please make it start from 1.

File: src/reanimated2/layoutReanimation/animationBuilder/commonTypes.ts [hunk 3]
```diff
@@ -60,13 +60,20 @@ export interface LayoutAnimationsValues {
   windowHeight: number;
 }
 
+export enum LayoutAnimationType {
+  ENTERING = 0,
```
Comment (tomekzaw) [inline src/reanimated2/layoutReanimation/animationBuilder/commonTypes.ts:64]: Please make it start from 1.

File: ios/native/NativeProxy.mm [hunk 4]
```diff
@@ -297,31 +298,30 @@ static CFTimeInterval calculateTimestampWithSlowAnimations(CFTimeInterval curren
...
-        yogaValues.setProperty(rt, [key UTF8String], matrix);
-      } else {
-        yogaValues.setProperty(rt, [key UTF8String], [(NSNumber *)value doubleValue]);
-      }
-    }
 
-    weakModule.lock()->layoutAnimationsManager().startLayoutAnimation(
-        rt, [tag intValue], std::string([type UTF8String]), yogaValues);
-  }];
+        weakModule.lock()->layoutAnimationsManager().startLayoutAnimation(rt, [tag intValue], type, yogaValues);
```
Comment (tomekzaw) [inline ios/native/NativeProxy.mm:320]: I don't think it's safe to use `weakModule.lock()->layoutAnimationsManager()`. If the module no longer exists, `.lock()` returns `nullptr` so we have a null pointer dereference. Can you please refactor all occurrences of `weakModule.lock()->layoutAnimationsManager()` into
```cpp
if (auto module = weakModule.lock()) {
  module->layoutAnimationsManager() // ...
}
```

**Review by tomekzaw — APPROVED** [review]

**Review by piaskowyk — APPROVED** [review]

**Pull request merged**